ddl/index.go

```diff
@@ -33,32 +34,71 @@ import (
 	"github.com/pingcap/tidb/util/types"
 )
 
-const maxPrefixLength = 767
+const maxPrefixLength = 3072
 
-func buildIndexInfo(tblInfo *model.TableInfo, unique bool, indexName model.CIStr,
-	idxColNames []*ast.IndexColName) (*model.IndexInfo, error) {
-	// build offsets
+func buildIndexColumns(columns []*model.ColumnInfo, idxColNames []*ast.IndexColName) ([]*model.IndexColumn, error) {
+	// build offsets.
 	idxColumns := make([]*model.IndexColumn, 0, len(idxColNames))
+
+	// the sum of length of all index columns.
+	sumLength := 0
+
 	for _, ic := range idxColNames {
-		col := findCol(tblInfo.Columns, ic.Column.Name.O)
+		col := findCol(columns, ic.Column.Name.O)
 		if col == nil {
 			return nil, errKeyColumnDoesNotExits.Gen("column does not exist: %s",
 				ic.Column.Name)
 		}
 
 		// Length must be specified for BLOB and TEXT column indexes.
 		if types.IsTypeBlob(col.FieldType.Tp) && ic.Length == types.UnspecifiedLength {
-			return nil, errors.Trace(errBlobKeyWithoutLength)
+			return nil, errBlobKeyWithoutLength
 		}
 
 		if ic.Length != types.UnspecifiedLength &&
-			!types.IsTypeChar(col.FieldType.Tp) &&
-			!types.IsTypeBlob(col.FieldType.Tp) {
-			return nil, errors.Trace(errIncorrectPrefixKey)
+			!types.IsTypeSpecifiable(col.FieldType.Tp) {
```

Comment:
Hi @zimulala and @coocood , please review it again, I fixed all style pitfalls besides the test mentioned by @zimulala .   I think this test isn't just for the bit type but also for varchar. It can be added lately